### PR TITLE
Late summer cleaning

### PR DESCRIPTION
**Pull request by natebrennand**
Adds a lot of comments.
The only logic change is adding a timeout for new jobs in the case that a packet is lost when a job starts.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -39,32 +46,39 @@ type Client struct {
...
+
+	// block while the client waits for confirmation that a job has been created
 	c.partialJobs <- &partialJob{data: data, warnings: warnings}
-	return <-c.newJobs, nil
+
+	// wait for confirmation, potentially times out
+	select {
+	case j := <-c.newJobs:
+		return j, nil
+	case <-time.After(10 * time.Second):
```
Comment (azylman) [inline client.go]: I'm pretty leery of adding this timeout when we don't know how long this usually takes. I imagine it's less than 10 seconds, but we've also never seen problems from this, so it seems more risky than useful.

File: client.go [hunk 2]
```diff
@@ -29,8 +30,14 @@ func (c noOpCloser) Close() error {
 
 var discard = noOpCloser{w: ioutil.Discard}
 
+type partialJob struct {
+	data     io.WriteCloser
+	warnings io.WriteCloser
```
Comment (azylman) [inline client.go]: Why put these on separate lines?

File: client.go [hunk 3]
```diff
@@ -80,54 +94,80 @@ func (c *Client) SubmitBackground(fn string, payload []byte) error {
...
 }
 
+// read attempts to read incoming packets from the gearman server to route them to the job
+// they are intended for.
 func (c *Client) read(scanner *bufio.Scanner) {
 	for scanner.Scan() {
 		pack := &packet.Packet{}
 		if err := pack.UnmarshalBinary(scanner.Bytes()); err != nil {
-			fmt.Printf("ERROR PARSING PACKET! %#v\n", err)
+			fmt.Printf("GEARMAN WARNING: error parsing packet! %#v\n", err)
```
Comment (azylman) [inline client.go]: We should probably print this to `os.Stderr`

File: client.go [hunk 4]
```diff
@@ -80,54 +94,80 @@ func (c *Client) SubmitBackground(fn string, payload []byte) error {
...
 		} else {
 			c.packets <- pack
 		}
 	}
 	if scanner.Err() != nil {
-		fmt.Printf("ERROR SCANNING! %#v\n", scanner.Err())
+		fmt.Printf("GEARMAN WARNING: error scanning! %#v\n", scanner.Err())
```
Comment (azylman) [inline client.go]: same comment about stderr

File: client.go [hunk 5]
```diff
@@ -80,54 +94,80 @@ func (c *Client) SubmitBackground(fn string, payload []byte) error {
...
 	}
 }
 
+// routePackets forwards incoming packets to the correct job.
 func (c *Client) routePackets() {
+	// operate on every packet that has been read
 	for pack := range c.packets {
+		if len(pack.Arguments) == 0 {
+			fmt.Println("GEARMAN WARNING: packet read with no handle!")
```
Comment (azylman) [inline client.go]: same comment about stderr

File: client.go (hunk 1 above)
Comment (natebrennand) [inline client.go]: agreed, I'm not convinced that this is the right move, but it seems really odd that the client is entirely blocked while waiting for a response.

File: client.go [hunk 6]
```diff
@@ -80,54 +94,80 @@ func (c *Client) SubmitBackground(fn string, payload []byte) error {
...
 			}()
-		} else {
-			c.getJob(handle) <- pack
+		default:
+			// send the packet to the right job
+			pktStream := c.getJob(handle)
+			if pktStream != nil {
+				pktStream <- pack
+			} else {
+				fmt.Printf("GEARMAN WARNING: packet read with handle of '%s', "+
```
Comment (azylman) [inline client.go]: same comment about stderr

File: client.go [hunk 7]
```diff
@@ -80,54 +94,80 @@ func (c *Client) SubmitBackground(fn string, payload []byte) error {
...
 				defer close(packets)
 				defer c.deleteJob(handle)
 				j.Run()
 			}()
-		} else {
-			c.getJob(handle) <- pack
+		default:
+			// send the packet to the right job
+			pktStream := c.getJob(handle)
+			if pktStream != nil {
```
Comment (azylman) [inline client.go]: good idea

File: client.go (hunk 2 above)
Comment (natebrennand) [inline client.go]: the terseness didn't seem to add any value, just more explicit

File: job/job.go [hunk 8]
```diff
@@ -63,17 +70,25 @@ func (j *Job) Run() State {
...
 
-func (j *Job) handlePackets(packets chan *packet.Packet) {
+// handlePackets updates a job based off of incoming packets associated with this job.
+func (j *Job) handlePackets(packets <-chan *packet.Packet) {
 	for pack := range packets {
 		switch pack.Type {
 		case packet.WorkStatus:
+			// check that packet is valid WORK_STATUS
+			if len(pack.Arguments) != 3 {
+				fmt.Printf("GEARMAN WARNING: Recieved invalid WORK_STATUS packet with '%d' fields\n",
```
Comment (azylman) [inline job/job.go]: same comment about stderr

File: job/job.go [hunk 9]
```diff
@@ -63,17 +70,25 @@ func (j *Job) Run() State {
...
+					len(pack.Arguments))
+				return
+			}
+
 			num, err := strconv.Atoi(string(pack.Arguments[1]))
 			if err != nil {
-				fmt.Println("Error converting numerator", err)
+				fmt.Println("GEARMAN WARNING: Error converting numerator", err)
```
Comment (azylman) [inline job/job.go]: same comment about stderr

File: job/job.go [hunk 10]
```diff
@@ -63,17 +70,25 @@ func (j *Job) Run() State {
...
 			}
 			den, err := strconv.Atoi(string(pack.Arguments[2]))
 			if err != nil {
-				fmt.Println("Error converting denominator", err)
+				fmt.Println("GEARMAN WARNING: Error converting denominator", err)
```
Comment (azylman) [inline job/job.go]: stderr

File: job/job.go [hunk 11]
```diff
@@ -84,14 +99,16 @@ func (j *Job) handlePackets(packets chan *packet.Packet) {
 			close(j.done)
 		case packet.WorkData:
 			if _, err := j.data.Write(pack.Arguments[1]); err != nil {
-				fmt.Printf("Error writing data, arg: %s, err: %s", pack.Arguments[1], err)
+				fmt.Printf("GEARMAN WARNING: Error writing data, arg: %s, err: %s",
```
Comment (azylman) [inline job/job.go]: stderr

File: job/job.go [hunk 12]
```diff
@@ -84,14 +99,16 @@ func (j *Job) handlePackets(packets chan *packet.Packet) {
...
+					pack.Arguments[1], err)
 			}
 		case packet.WorkWarning:
 			if _, err := j.warnings.Write(pack.Arguments[1]); err != nil {
-				fmt.Printf("Error writing warnings, arg: %s, err: %s", pack.Arguments[1], err)
+				fmt.Printf("GEARMAN WARNING: Error writing warnings, arg: %s, err: %s",
```
Comment (azylman) [inline job/job.go]: stderr

File: job/job.go [hunk 13]
```diff
@@ -84,14 +99,16 @@ func (j *Job) handlePackets(packets chan *packet.Packet) {
...
+					pack.Arguments[1], err)
 			}
 		default:
-			fmt.Println("WARNING: Unimplemented packet type", pack.Type)
+			fmt.Println("GEARMAN WARNING: Unimplemented packet type", pack.Type)
```
Comment (azylman) [inline job/job.go]: stderr

File: scanner/scanner.go [hunk 14]
```diff
@@ -7,6 +7,9 @@ import (
 	"io"
 )
 
+// needMoreData returns the requisite values to get the bufio.Scanner to send more data to
```
Comment (azylman) [inline scanner/scanner.go]: nice

File: client.go (hunk 1 above)
Comment (azylman) [inline client.go]: The client isn't blocked, the goroutine that you're calling `Submit` in is blocked, which isn't the same thing - all of the processing of packets for existing jobs happen in different goroutines

File: client.go (hunk 2 above)
Comment (azylman) [inline client.go]: It's a nit, but I'm pretty sure the other one is considered more idiomatic Go. It'd make sense for them to be on separate lines if they were each getting their own comment, but when they're not, it just seems unnecessary to me

Comment (natebrennand) [inline client.go]: adding comments

File: client.go (hunk 1 above)
Comment (natebrennand) [inline client.go]: Sorry, I meant blocked from submitting more jobs.

Comment (azylman) [inline client.go]: That's intentional, in my mind - if you want to submit jobs in the background, you can do it in a goroutine

**Comment by azylman** [comment]
lgtm

**Pull request merged**